emacs/emacs_config.org
itle: Emacs Configuration
#+PROPERTY: header-Ares :results none :tangle ~/.emacs.d/init.el
#+auto_tangle: t


* About this config
This is my first attempt at a fully-feature emacs configuration. In many ways, I have little idea of what I'm doing. I've used emacs for years without these features, but with the advent of LSP would like to learn more about to make emacs work the best for my workflow as possible. I've learned a lot via DOOM emacs, but now I want to ensure I have a system that caters to my personal need.

* Setup:
I got tired of emacs asking if it was ok to follow a symliink on every boot and the attempts I made to suppress that didn't work. So instead, we tangle directly to =init.el= and save this org file under version control.

** Define where we go for packages

** Straight.el
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 6))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
#+end_src

#+RESULTS:
: t

** Setup use-package

#+BEGIN_SRC emacs-lisp :results none :tangle ~/.emacs.d/init.el
  ;;Initalize use-package
  (unless (package-installed-p 'use-package)
  (package-install 'use-package))

  (require 'use-package)



#+END_SRC
* Initial Quick setup
These are the random one-off configurations that help make Emacs better, but didn't fall into a great category.  Things like removing the unneeded ui elements from the screen or suppressing annoying dialogs that you always answer yes to.
** Remove some of the features of the UI that we don't need (toolbars, menubars, scroll bars, and splash screen)
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
  ;;-*- lexical-binding: t; -*-
  ;; Don't show the splash screen
  (setq inhibit-startup-message t)
  (setq initial-buffer-choice nil)
  ;; Tune garbage collection
  (setq gc-cons-threshold (* 50 1000 1000))

  ;; Turn off some unneeded UI Elements
  (menu-bar-mode -1)
  (tool-bar-mode -1)
  (scroll-bar-mode -1)
  (display-time-mode 1)
#+END_SRC

#+RESULTS:
: t

** Make deleted files go to trash
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
  (setq delete-by-moving-to-trash t
trash-directory "~/.local/share/Trash/files/")
#+end_src

#+RESULTS:
: ~/.local/share/Trash/files/

** Turn on beacon
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
  (use-package beacon
  :straight (beacon :fetcher github :repo "Malabarba/beacon")
    :ensure t
    :config
    (beacon-mode +1))
#+end_src

#+RESULTS:
: t

** DONE Suppress execute confirm
CLOSED: [2022-12-28 Wed 17:26]
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
 (setq org-confirm-babel-evaluate nil)
#+end_src

#+RESULTS:

** DONE Keep little files from littering
CLOSED: [2022-12-28 Wed 18:13]

** DONE Suppress compiler warnings
CLOSED: [2022-12-28 Wed 17:27]
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 

;; Silence compiler warnings as they can be pretty disruptive
(setq native-comp-async-report-warnings-errors nil)

;; Set the right directory to store the native comp cache
(add-to-list 'native-comp-eln-load-path (expand-file-name "eln-cache/" user-emacs-directory))

#+end_src

#+RESULTS:
| /Users/rcool2/.emacs.d/eln-cache/ | /opt/homebrew/Cellar/emacs-plus@30/30.0.50/Emacs.app/Contents/native-lisp/ |

** Load custom lisp files
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
  (load (expand-file-name "rc-funcs.el" user-emacs-directory))
  (setq custom-file (expand-file-name "custom.el" user-emacs-directory))
  (load custom-file)
#+end_src

#+RESULTS:
: t

** Suppress emacs from asking if it should follow symlink files
#+begin_src emacs-lisp  :tangle ~/.emacs.d/init.el
(setq vc-follow-symlinks t)
(setq vc-suppress-confirm t)
#+end_src

#+RESULTS:
: t

** Turn on line numbers
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
  ;; Turn on line numbers in every buffer
  (global-display-line-numbers-mode 1)
  ;; With vim bindings, relatives let us use j and k better
  (setq display-line-numbers-type 'relative)
#+END_SRC

#+RESULTS:
: relative

** Disable lockfiles
#+begin_src emacs-lisp  :tangle ~/.emacs.d/init.el
(setq create-lockfiles nil)
#+end_src

#+RESULTS:

** Line Highlighting and Fringe
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
  ;; Turn on line highlithting for current line
  (hl-line-mode 1)
  ;; Add some margins
  (set-fringe-mode 10)
#+END_SRC

#+RESULTS:
: ((vertical-scroll-bars) (left-fringe . 10) (right-fringe . 10))

** Add escape to kill prompts
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
  ;; Make escape kill prompts
  (global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+END_SRC

#+RESULTS:
: keyboard-escape-quit

** Keys for Mac
#+begin_src emacs-lisp  :tangle ~/.emacs.d/init.el
  (setq mac-command-modifier       'meta
        mac-option-modifier        'meta
        mac-control-modifier       'control
        mac-right-option-modifier     'meta
        mac-right-control-modifier  'control)
#+end_src

#+RESULTS:
: control

** DONE Turn on garbage collection
CLOSED: [2022-12-28 Wed 18:13]
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package gcmh
    :diminish gcmh-mode
    :straight (gcmh-mode :host github :repo "akirak/gcmh")
    :config
    (setq gcmh-idle-delay 5
	  gcmh-high-cons-threshold (* 16 1024 1024))  ; 16mb
    (gcmh-mode 1))

  (add-hook 'emacs-startup-hook
	    (lambda ()
	      (setq gc-cons-percentage 0.1))) ;; Default value for `gc-cons-percentage'

  (add-hook 'emacs-startup-hook
	    (lambda ()
	      (message "Emacs ready in %s with %d garbage collections."
		       (format "%.2f seconds"
			       (float-time
				(time-subtract after-init-time before-init-time)))
		       gcs-done)))


#+end_src

#+RESULTS:
| lambda  | nil | (message Emacs ready in %s with %d garbage collections. (format %.2f seconds (float-time (time-subtract after-init-time before-init-time))) gcs-done) |                                                                                  |
| lambda  | nil | (setq gc-cons-percentage 0.1)                                                    |                                                                                  |
| closure | (t) | nil                                                                              | (message Emacs ready in %s with %d garbage collections. (format %.2f seconds (float-time (time-subtract after-init-time before-init-time))) gcs-done) |
| closure | (t) | nil                                                                              | (setq gc-cons-percentage 0.1)                                                    |

** Manage how new windows are placed
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (winner-mode +1)
  (setq display-buffer-base-action
    '((display-buffer-reuse-window
       display-buffer-reuse-mode-window
       display-buffer-same-window
       display-buffer-in-previous-window)))
#+end_src

#+RESULTS:: t


* Build the UI we want
** Install a better mode line

#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
    (use-package doom-modeline
  :straight (doom-modeline :repo "seagle0128/doom-modeline" :fetcher github)
      :ensure t
      :init (doom-modeline-mode 1)
      :config
      (setq doom-modeline-hud nil)
      (setq doom-modeline-minor-modes nil)
      (setq rc-doom-modeline-text-height 200)
      (setq doom-modeline-height 35)
      )
    (use-package all-the-icons
:straight (all-the-icons
 :repo "domtronn/all-the-icons.el"
 :fetcher github
 :files (:defaults "data"))
      :ensure t
      :if (display-graphic-p))

#+END_SRC

#+RESULTS:
: t

** Themes:
*** Setup Modus Themes

I like to do this even if I don't plan on the modus themes being my main theme of choice. I often swap themes, and this way they are ready when I want them.
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
  (use-package modus-themes
  :ensure t
  :straight (modus-themes :fetcher sourcehut :repo "protesilaos/modus-themes")
  :init
  (setq modus-themes-hl-line '(intense)
  modus-themes-subtle-line-numbers nil
  modus-themes-region '(no-extend bg-only)
  modus-themes-variable-pitch-ui nil
  modus-themes-fringes 'intense
  modus-themes-diffs nil
  modus-themes-italic-constructs t
  modus-themes-bold-constructs  t
  modus-themes-prompts '( bold intense italic)
  modus-themes-intense-mouseovers t
  modus-themes-paren-match '(bold intense)
  modus-themes-syntax '( yellow-comments)
  modus-themes-links '(neutral-underline background)
  modus-themes-mode-line '(moody borderless accented 4 1)
  modus-themes-headings
   '((1 . (variable-pitch 1.1 rainbow))
                         (2 . (regular))
                         (3 . (regular))
                         (4 . (regular))
                         (t . (rainbow)))
  modus-themes-org-blocks `gray-background
   ))




#+END_SRC

** Mixed Pitch
A better verson of variable pitch.

#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
  (use-package mixed-pitch
    :ensure t
    :straight (mixed-pitch :fetcher gitlab :repo "jabranham/mixed-pitch")
    :defer t
    :config
    (setq mixed-pitch-set-height nil)
    (dolist (face '(org-date org-priority org-tag org-special-keyword))
      (add-to-list 'mixed-pitch-fixed-pitch-faces face))
   )
#+end_src

#+RESULTS:

*** doom-themes
#+begin_src emacs-lisp  :tangle ~/.emacs.d/init.el
    (use-package doom-themes
    :straight t
    :after mixed-pitch
      :ensure t
      :config
      (setq doom-themes-enable-bold t
      doom-themes-enable-italic t)
      (load-theme 'doom-1337 t)

      (doom-themes-visual-bell-config)
      (doom-themes-neotree-config)
      (setq doom-themes-treemacs-theme "doom-atom")
      (doom-themes-treemacs-config)
      (doom-themes-org-config)
  )
  
    (load-theme 'doom-Iosvkem t)
  #+end_src

  #+RESULTS:

*** Kaolin Themes
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package kaolin-themes 
    :straight (kaolin-themes
 :fetcher github
 :repo "ogdenwebb/emacs-kaolin-themes"
 :old-names (kaolin-theme)
 :files (:defaults "themes/*.el"))
      :ensure t
      :config
      (setq kaolin-themes-modeline-border nil)
  )
#+end_src
*** Ef-themes
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package ef-themes
      :ensure t
      :straight (ef-themes :host github :repo "protesilaos/ef-themes")
      :init
      (setq ef-themes-headings
	    (quote ((1 . (variable-pitch 1.1))
		    (2 . (regular))
		    (3 . (regular))
		    (4 . (regular))))))
#+end_src



** Setup our fonts
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
   (set-face-attribute 'default nil :family "PragmataProMonoLiga Nerd Font" :height 200)

   ;; Proportionally spaced typeface
   (set-face-attribute 'variable-pitch nil :family "Anonymice Nerd Font Mono" :height 1.0)

   ;; Monospaced typeface
   (set-face-attribute 'fixed-pitch nil :family "Spleen 32x64" :height 1.5)


   (if (facep 'mode-line-active)
       (set-face-attribute 'mode-line-active nil :family "Spleen 32x64" :height 200) ; For 29+
     (set-face-attribute 'mode-line nil :family "Spleen 32x64" :height 200))
  (set-face-attribute 'mode-line-inactive nil :family "Anonymice Nerd Font Mono" :height 200)
#+END_SRC

#+RESULTS:

** Update the help docs

#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
  (use-package helpful
  :straight (helpful :repo "Wilfred/helpful" :fetcher github)

    :ensure t)
  (global-set-key (kbd "C-c C-d") #'helpful-at-point)
  (global-set-key (kbd "C-h F") #'helpful-function)

  (global-set-key (kbd "C-h C") #'helpful-command)

#+END_SRC
** Popper.el - tame popups
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
  (use-package popper
  :straight (popper :fetcher github :repo "karthink/popper")
    :ensure t
    :bind (("C-`" . popper-toggle-latest) 
           ("M-`" . popper-cycle)
           ("C-M-`" . popper-toggle-type))
    :init
    (setq popper-reference-buffers
          '("\\*Messages\\*"
            "Output\\*$"
            "\\*Warnings\\*"
            "\\*Backtrace\\*"
            help-mode
            helpful-mode
            compilation-mode))
    (popper-mode +1)
    (popper-echo-mode +1)
    )


  (defvar rcool/help-temp-buffers '("^\\*Help\\*$",
                                    "^\\*helpful command"))

  (while rcool/help-temp-buffers
    (add-to-list 'display-buffer-alist
                 `(,(car rcool/help-temp-buffers)
                   (display-buffer-in-side-window
                    (side . left)
                    (window-width . 50))))

    (setq rcool/help-temp-buffers (cdr rcool/help-temp-buffers)))
#+end_src

** Key bindings with general
#+begin_src emacs-lisp  :tangle ~/.emacs.d/init.el
  (use-package general
    :ensure t
    :straight (general
 :fetcher github
 :repo "noctuid/general.el")
    :config
    (general-auto-unbind-keys)
    (general-evil-setup t))



  #+end_src

** Ace window
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package ace-window 
    :straight (ace-window :repo "abo-abo/ace-window"
            :fetcher github)
    :defer t :ensure t)
#+end_src
** Diminish
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package diminish 
:straight (diminish :fetcher github :repo "myrjola/diminish.el")
    :ensure t)
#+end_src
** Bufler 
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package bufler
      :ensure t
      :straight (bufler :fetcher github :repo "alphapapa/bufler.el"
        :files (:defaults (:exclude "helm-bufler.el")))
      :config
      (setq bufler-filter-buffer-modes nil
           bufler-filter-buffer-name-regexps nil)
      (setf bufler-groups (bufler-defgroups
                           (group
                            ;; All named workspaces
                            (auto-workspace))
                           (group
                            ;; `help-mode' and `info-mode'
                            (group-or "*Help/Info*"
                                      (mode-match "*Help*" (rx bos "help-"))
                                      (mode-match "*Info*" (rx bos "info-"))))
                           (group
                            ;;special buffers (except magit)
                            (group-and "*Special*"
                                       (lambda (buffer)
                                         (unless (or (funcall (mode-match "Magit" (rx bos "magit-status"))
                                                              buffer)
                                                     (funcall (mode-match "Dired" (rx bos "dired"))
                                                              buffer)
                                                     (funcall (auto-file) buffer))
                                           "*Special*"))))
                           (group
                            ;;Special Special Buffers
                            (name-match "**Special**"
                                        (rx bos "*" (or "Messages" "Warnings" "scratch" "Backtrace") "*")))
                           (group
                            ;; All other Magic buffers by directory
                            (mode-match "Magit* (non-status)" (rx bos (or "magit" "forge") "-"))
                            (auto-directory))
                           ;; Remaining by directory then major mode
                           (auto-directory)
                           (auto-mode)))
      :general
      (:keymaps 'bufler-list-mode-map "Q" 'kill-this-buffer))
#+end_src

* Developer Interaction
This class of settings I think of as those that I need in order to work best with the editor.  How can I commicate. Plugins that affect how I communicate via keybindings, mouse, whatever should go here.
** Avy
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
(use-package avy :ensure t :straight (avy :repo "abo-abo/avy"
     :fetcher github))
#+end_src

** Undo-fu
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
(use-package undo-fu :ensure t :straight (undo-fu :fetcher codeberg :repo "ideasman42/emacs-undo-fu"))
#+end_src
** flycheck
#+begin_src emacs-lisp :tangle ~/.emacs.d/init.el 
    (use-package flycheck :straight (flycheck :repo "flycheck/flycheck" :fetcher github)
  :ensure t
    :config
  (global-flycheck-mode))

  (use-package exec-path-from-shell :straight (exec-path-from-shell :fetcher github :repo "purcell/exec-path-from-shell")
  :ensure t
  :config
  (exec-path-from-shell-initialize))
  
#+end_src



** Evil
#+BEGIN_SRC emacs-lisp  :tangle ~/.emacs.d/init.el
    (defun modeline-face-color ()
      (let ((color (cond ((minibufferp) '("#1d1d1f" . "#ffffff"))
			 ((evil-insert-state-p) '("#331612" . "#ffffff"))
			 ((evil-emacs-state-p) '("#1a2b14" . "#ffffff"))
			 ((evil-visual-state-p) '("#5a5c21" . "#ffffff"))
			 ((evil-replace-state-p) '("#1a403e" . "#ffffff"))
			 ((evil-motion-state-p) '("#677691" . "#ffffff"))
			 ((buffer-modified-p) '("#2b1a40" . "#ffffff"))
			 (t '("#1d1d1f" . "#ffffff")))))
	(set-face-background 'mode-line (car color))
	(set-face-foreground 'mode-line (cdr color))))
    (use-package evil
      :ensure t
      :straight (evil :repo "emacs-evil/evil"
	:fetcher github
	:files (:defaults
		"doc/build/texinfo/evil.texi"
		(:exclude "evil-test-helpers.el")))
      :init
      (setq evil-want-fine-under t)
      (setq evil-want-keybinding nil)
      (setq evil-want-Y-yank-to-eol t)
      (setq evil-undo-system 'undo-fu)
      (setq evil-search-module 'evil-search)
      :config

      (evil-set-initial-state 'dashboard-mode 'motion)
      (evil-set-initial-state 'debugger-mode 'motion)
      (evil-set-initial-state 'pdf-view-mode 'motion)
      (evil-set-initial-state 'bufler-list-mode 'emacs)
      (evil-set-initial-state 'inferior-python-mode 'emacs)
      (evil-set-initial-state 'term-mode 'emacs)
      (evil-set-initial-state 'calc-mode 'emacs)
      (define-key evil-window-map "\C-q" 'evil-delete-buffer) ;; Maps C-w C-q to evil-delete-buffer (The first C-w puts you into evil-window-map)
      (define-key evil-window-map "\C-w" 'kill-this-buffer)
      (define-key evil-motion-state-map "\C-b" 'evil-scroll-up) ;; Makes C-b how C-u is
      (add-hook 'post-command-hook 'modeline-face-color)

      ;; ----- Setting cursor colors
      (setq evil-emacs-state-cursor    '("#649bce" box))
      (setq evil-normal-state-cursor   '("#d9a871" box))
      (setq evil-operator-state-cursor '("#ebcb8b" hollow))
      (setq evil-visual-state-cursor   '("#677691" box))
      (setq evil-insert-state-cursor   '("#eb998b" (bar . 2)))
      (setq evil-replace-state-cursor  '("#eb998b" hbar))
      (setq evil-motion-state-cursor   '("#ad8beb" box))
  (with-eval-after-load 'copilot
    (evil-define-key 'insert copilot-mode-map
		     (kbd "<tab>") #'my/copilot-tab))
      (evil-mode +1))
    (use-package evil-collection
      :ensure t
      :after evil
      :straight

.config/doom/config.el
(setq mac-command-modifier       'meta
      mac-option-modifier        'meta
      mac-control-modifier       'control
      mac-right-option-modifier     'meta
      mac-right-control-modifier  'control)

(set-frame-parameter nil 'alpha-background 50)

(use-package! company
  :bind
  (:map company-active-map
        ("<tab>" . company-complete-selection))
  :custom
  (company-minimum-prefix-length 1)
  (company-idl-delay 0.0))

(setq user-full-name "Richard Cool"
      user-mail-address "richardjcool@gmail.com")

(setq-default delete-by-moving-to-trash t
              trash-directory "~/.local/share/Trash/files/")

(setq-default window-combination-resize t)

;; (defadvice! prompt-for-buffer (&rest _)
;;   :after '(evil-window-split evil-window-vsplit)
;;   (counsel-ibuffer))

(after! undo-fu
  (setq undo-limit 10000000 ;; 1MB
        undo-strong-limit 100000000 ;;100MB
        undo-outer-limit 1000000000) ;; 1GB
(setq undo-fu-allow-undo-in-region t
      undo-fu-ignore-keyboard-quit t))
;;Evil undo
(after! evil
  (setq evil-want-fine-undo t))

(when (daemonp)
  (add-hook! '(delete-frame-functions delete-terminal-functions)
             (let ((inhibit-message t))
               (recentf-save-list)
               (savehist-save))))

(setq doom-font (font-spec :family "PragmataProMonoLiga Nerd Font" :size 18)
      doom-big-font (font-spec :family "Iosevka Aile" :size 30)
      doom-variable-pitch-font (font-spec :family "Iosevka Aile" :size 18)
      doom-unicode-font (font-spec :family "Spleen 32x64" :size 18)
      doom-serif-font (font-spec :family "Iosevka Aile" :size 18)

      )


(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

;; accept completion from copilot and fallback to company
(use-package! copilot
  :hook (prog-mode . copilot-mode)
  :bind (("C-TAB" . 'copilot-accept-completion-by-word)
         ("C-<tab>" . 'copilot-accept-completion-by-word)
         :map copilot-completion-map
         ("<tab>" . 'copilot-accept-completion)
         ("TAB" . 'copilot-accept-completion)))

(define-globalized-minor-mode global-rainbow-mode rainbow-mode
  (lambda ()
    (when (not (memq major-mode
                (list 'org-agenda-mode)))
     (rainbow-mode 1))))
(global-rainbow-mode 1 )

(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)

(setq tramp-default-method "ssh")

(use-package! org-auto-mode
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config
  (setq org-auto-tangle-default t))

(after! org

  ;; TODO: extract org-directory into noweb reference so it can be used in all the places
  (setq org-directory "~/org"
        org-agenda-files '(
                           "~/org/todo.org"
                           "~/org/agenda.org"
                           )
        +org-capture-todo-file "inbox.org")
  (setq org-todo-keywords '((sequence "TODO(t)" "WAIT(w)" "HOLD(h)" "|" "DONE(d!)" "CANCELLED(c!)")))

  (setq org-todo-keyword-faces '(("TODO" . (:foreground "#7bc275" :weight normal))
                                 ("WAIT" . (:foreground "orange" :weight normal))
                                 ("HOLD" . (:inherit warning :weight normal))))
  (setq org-tag-keywords '((sequence "home (h)" "work (w)" "qoedash(q)" "sessionwiz(s)" "deviceReachDash(d)" "adhoc(a)")))

  (org-babel-do-load-languages
   'org-babel-load-languages
   '(
     (python . t)
     (shell . t)
     (js . t)
     (lua . t)
     (sql . t)
     ))
  (setq python-shell-completion-native-enable nil)
  (setq org-src-window-setup 'current-window)

   (defun org-babel-execute:typescript (body params)
    (let ((org-babel-js-cmd "npx ts-node < "))
      (org-babel-execute:js body params)))
(setq org-src-fontify-natively t
      org-src-tab-acts-natively t
      org-src-window-setup 'current-window)

  (setq org-capture-templates
        '(
          ("t" "todo" entry (file +org-capture-todo-file) "* TODO %?")
          ("n" "node" entry (file +org-capture-todo-file) "* Note: ")
          ("p" "process email" entry (file +org-capture-todo-file)
           "* TODO %? %:fromname: %a")
          )))
(setq org-startup-indented t
      org-pretty-entities t
      org-hide-emphasis-markers t
      org-startup-with-inline-images t
      org-image-actual-width '(300))

(use-package! org-appear
  :hook (org-mode . org-appear-mode))

(use-package! org-superstar
  :config
  (setq org-superstar-special-todo-items t)
  (add-hook 'org-mode-hook (lambda() (org-superstar-mode +1))))
(setq-default line-spacing 6)
(use-package! olivetti
  :init
  (setq olivetti-body-width 0.67)
  :config
  (defun distraction-free ()
    "Distraction free writing environment"
    (interactive)
    (if (equal olivetti-mode nil)
        (progn
          (window-configuration-to-register 1)
          (delete-other-windows)
          (text-scale-increase 2)
          (olivetti-mode t))
      (progn
        (jump-to-register 1)
        (olivetti-mode 0)
        (text-scale-decrease 2))))
  :bind
  (("<f9>" . distraction-free)))

(map!
 :leader
 :prefix "m"
 :map 'org-src-mode-map
 (:desc "Accept SRC Edits"
 :nmv "J" #'org-edit-src-exit)
 (:desc "Abort SRC Edits"
 :nmv "X" #'org-edit-src-abort))

(setq org-hugo-base-dir "~/org/markdown")

(use-package! beacon
  :config
  (beacon-mode +1))

;; Turn on line highlithting for current line
(hl-line-mode 1)
;; Add some margins
(set-fringe-mode 10)

(winner-mode +1)
(setq display-buffer-base-action
  '((display-buffer-reuse-window
     display-buffer-reuse-mode-window
     display-buffer-same-window
     display-buffer-in-previous-window)))

(after! doom-modeline
    (setq
     doom-modeline-hud t
     doom-modeline-minor-modes nil
          doom-modeline-height 25))

(use-package! mixed-pitch
  :defer t
  :hook (text-mode . mixed-pitch-mode)
  :config
  (set-face-attribute 'default nil :font "Spleen 32x64" :height 130)
  (set-face-attribute 'fixed-pitch nil :font "Iosekva Term")
  (set-face-attribute 'variable-pitch nil :font "Iosevka Aile"))
 (add-hook 'mixed-pitch-mode-hook #'solaire-mode-reset)

(use-package! quickrun
   :defer t
   :general
   (general-def
    :states 'normal
    :prefix "SPC"
    :keymaps 'quickrun--mode-map
    "cq" '(nil :which-key "quickrun")
    "cqq" '(quit-window :which-key "Quit")
    "cqr" '(quickrun :which-key "Run")
    "cqR" '(quickrun-region :which-key "Run Region")
    "cqa" '(quickrun-with-arg :which-key "Run with [A]rgs")
    "cqm" '(quickrun-autorun-mode :which-key "Toggle autorun mode")
    "cqs" '(quickrun-select :which-key "Select backend")
    "cq"  '(nil :which-key "quickrun")
    "cqq" '(quit-window :which-key "Quit")
    "cqr" '(quickrun :which-key "Run")
    "cqR" '(quickrun-region :which-key "Run Region")
    "cqa" '(quickrun-with-arg :which-key "Run with [A]rgs")
    "cqm" '(quickrun-autorun-mode :which-key "Toggle autorun mode")
    "cqs" '(quickrun-select :which-key "Select backend")
))

(after! tree-sitter
(add-to-list 'tree-sitter-major-mode-language-alist '(tsx-mode . tsx))
)

(use-package! lsp-ui
  :after lsp
  :hook ((lsp-mode . lsp-ui-mode)
         (lsp-mode . lsp-ui-sideline-mode))
  :config
  (setq lsp-ui-sideline-show-hover t)
  (setq lsp-ui-doc-position 'top)
  )

(use-package! tsi
  :hook ((web-mode-hook . tsi-typescript-mode)
         (typescript-mode-hook . tsi-typescript-mode)
         (tsx-mode-hook . tsi-typescript-mode)
          (json-mode-hook . tsi-typescript-mode)
          (css-mode-hook . tsi-css-mode)
          (scss-mode-hook . tsi-scss-mode))
  :config
  (require 'tsi-css)
  (require 'tsi-json)
  (require 'tsi-typescript)

)

(use-package! tsx-mode

  :hook (tsx-mode . lsp-deferred)
  :config
  (add-to-list 'auto-mode-alist '("\\.[jt]sx?\\'" . tsx-mode))

(map!
 :leader
 :prefix "m"
 :map 'general-override-mode-map
 (:desc "Toggle All Nodes"
  :nmv #'tsx-mode-fold-toggle-all-nodes)
 (:desc "Toggle Coverage"
        :nmv #'tsx-mode-coverage-toggle)
 (:desc "Toggle Node"
        :nvm #'tsx-mode-fold-toggle-node)
)
)

(use-package! apheleia
  :config
  (apheleia-global-mode +1))

(general-def
  :prefix-map 'rc/lsp-map
"d"   #'lsp-find-declaration
"D"   #'lsp-ui-peek-find-definitions
"R"   #'lsp-ui-peek-find-references
"i"   #'lsp-ui-peek-find-implementation
"t"   #'lsp-find-type-definition
"s"   #'lsp-signature-help
"o"   #'lsp-describe-thing-at-point
"r"   #'lsp-rename

"f"   #'lsp-format-buffer
"m"   #'lsp-ui-imenu
"x"   #'lsp-execute-code-action
)
(hercules-def
:toggle-funs #'rc/lsp-map-mode
:keymap 'rc/lsp-map
:transient t)
(map!
 :leader
 :prefix "H"
 :desc "LSP"
 :nm "L" #'lsp-map-mode
 )

(general-def
:prefix-map 'rc/flycheck-map
   "f" #'flycheck-error-list-set-filter
   "j" #'flycheck-next-error
   "k" #'flycheck-previous-error
)
(hercules-def
 :toggle-funs #'rc/flycheck-mode
 :keymap 'rc/flycheck-map
 :transient t)
(map!
 :leader
 :prefix "H"
 :desc "Flycheck"
 :nm "f" #'rc/flycheck-mode)

(general-def
  :prefix-map 'rc/avy-map
"c" #'avy-goto-char-timer
"C" #'avy-goto-char
"w" #'avy-goto-word-1
"W" #'avy-goto-word-0
"l" #'avy-goto-line
"L" #'avy-goto-end-of-line
"m" #'avy-move-line
"M" #'avy-move-region
"k" #'avy-kill-whole-line
"K" #'avy-kill-region
"y" #'avy-copy-line
"Y" #'avy-copy-region
)
(hercules-def
 :toggle-funs #'rc/avy-mode
 :keymap 'rc/avy-map
 :transient t
)
(map!
 :leader
 :prefix "H"
 :desc "Avy"
 :nm "a" #'rc/avi-mode)

;; (general-def
;;   :prefix-map 'rc/snippet-map
;;   "d" #'yas-load-directory
;;   "e" #'yas-activate-extra-mode
;;   "i" #'yas-insert-snippet
;;   "f" #'yas-visit-snippet-file
;;   "n" #'yas-new-snippet
;;   "t" #'yas-tryout-snippet
;;   "l" #'yas-describe-tables
;;   "g" #'yas/global-mode
;;   "m" #'yas/minor-mode
;;   "a" #'yas-reload-all
;; )
;; (hercules-def
;;  :toggle-funs #'rc/snippet-mode
;;  :keymap 'rc/snippet-map
;;  :transiet t)

;; (map!
;;  :leader
;;  :prefix "H"
;;  :desc "Snippet"
;;  :nm "s" #'rc/snippet-mode)

(general-def
  :prefix-map 'rc/origami-map
   "c" #'origami-close-node
   "n" #'origami-next-fold
   "p" #'origami-previous-fold
   "f" #'origami-forward-toggle-node
   "a" #'origami-toggle-all-nodes
   "s" #'origami-show-only-node
   )
(hercules-def
 :toggle-funs #'rc/origami-mode
 :keymap 'rc/origami-map
 :transient t)
(map!
:leader
:prefix "H"
:desc "Folding"
:nm "f" #'rc/origami-mode
)

(hercules-def
 :toggle-funs #'rc/magit-map
 :keymap 'magit-mode-map
 :transient t)
(hercules-def
 :toggle-funs #'rc/buffer-mode
 :keymap 'doom-leader-buffer-map
:transient t)

(general-def
  :prefix-map 'my-dired-map

  "\\" #'dired-do-ispell
  "(" #'dired-hide-details-mode
  ")" #'dired-omit-mode
  "+" #'dired-create-directory
  "=" #'diredp-ediff         ;; smart diff
  "?" #'dired-summary
  "$" #'diredp-hide-subdir-nomove
  "A" #'dired-do-find-regexp
  "C" #'dired-do-copy        ;; Copy all marked files
  "D" #'dired-do-delete
  "E" #'dired-mark-extension
  "e" #'dired-ediff-files
  "F" #'dired-do-find-marked-files
  "G" #'dired-do-chgrp
  "g" #'revert-buffer        ;; read all directories again (refresh)
  "i" #'dired-maybe-insert-subdir
  "l" #'dired-do-redisplay   ;; relist the marked or singel directory
  "M" #'dired-do-chmod
  "m" #'dired-mark
  "O" #'dired-display-file
  "o" #'dired-find-file-other-window
  "Q" #'dired-do-find-regexp-and-replace
  ;; "R" #'dired-do-rename
  ;; "r" #'dired-do-rsynch
  ;; "S" #'dired-do-symlink
  ;; "s" #'dired-sort-toggle-or-edit
  ;; "t" #'dired-toggle-marks
  ;; "U" #'dired-unmark-all-marks
  ;; "u" #'dired-unmark
  ;; "v" #'dired-view-file      ;; q to exit, s to search, = gets line #
  ;; "w" #'dired-kill-subdir
  ;; "Y" #'dired-do-relsymlink
  "z" #'diredp-compress-this-file
  "Z" #'dired-do-compress
  )

(hercules-def
 :toggle-funs #'my-dired-mode
 :keymap 'my-dired-map
 :transient t)
